dev/bots/count-github-tasks.py

```diff
@@ -0,0 +1,32 @@
+# Copyright 2019 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+# Script that reports how many GitHub tasks autorollers should expect
+# on PRs on this repository.
+#
+# This assumes all the non-skipped tasks run, which is not accurate
+# for PRs that don't affect the engine version!
+
+# This file must be in Python because it is used by the Skia
+# autoroller logic and that logic does not have Dart available.
+
+import re
+
+# First count the Cirrus tasks...
+cirrusYaml = open('.cirrus.yml', 'r').readlines()
```

Comment:
Could we instead download the .cirrus.yml file from https://raw.githubusercontent.com/flutter/engine/master/... and parse it?

That way the roller can, without a checkout, download this file and execute it.

Reply:
@Hixie

Reply:
The `.cirrus.yml` file that matters is the one in the same repo as the PR that you are submitting (so for the engine autoroller, it's the one in flutter/flutter, the same repo that you have to have checked out already so that you can change `engine.version`).

Reply:
I understand. I meant that making the script executable without having a checkout will make it easier to use from the autoroller because of the way it's structured - code review (gerrit/github) logic is somewhat separated from the parts that deal with the checkout.

This is why when we implemented reading the merge_method from flutter we do it from https://raw.githubusercontent.com/flutter/flutter/master/bin/internal/engine.merge_method
Config: https://skia.googlesource.com/buildbot/+/master/autoroll/config/flutter-engine-flutter.json#27

Reply:
One more reason it would be better to have the script be executable standalone:
We would like to eventually move Github rollers to be checkout-less (like we did for some Gerrit rollers- https://skia.googlesource.com/buildbot/+/master/autoroll/go/repo_manager/no_checkout_deps_repo_manager.go)
The roller would then use the Github API to create changes without using a local checkout. Will not be able to do this for engine.git rollers (because of the license script) but we might be able to do this for flutter.git rollers one day.

Reply:
The script would need to be given the exact commit being rolled so that it could go and find the .cirrus.yml file (and eventually LUCI config) in that commit. What API would you like? Full commit hash in first command line argument?

Reply:
Full commit hash in first command line argument SGTM. Thanks!

Reply:
Since this is for auto-roller only, and the auto-roller would never touch the `.cirrus.yaml` file, I think we can just use the tip-of-tree `.cirrus.yaml` to get the number?

Reply:
The roller uses this API to get the number of checks: https://skia.googlesource.com/buildbot/+/master/go/github/github.go#265
IIRC, it uses this because the checks API was not returning all checks (atleast when this was added in 7/2018).

I just noticed that for a engine->flutter roll from: https://autoroll.skia.org/r/flutter-engine-flutter-autoroll?tab=status
the roller thinks there are a total of 30 checks:

W1031 13:38:48.789797       1 roll.go:346] roll.go:346 len(tryResults) != checksNum: 30 != 12
I1031 13:38:48.790003       1 autoroll.go:209] autoroll.go:209   hostonly_devicelab_tests-1-linux #824634067824: SUCCESS (cq)
I1031 13:38:48.790130       1 autoroll.go:209] autoroll.go:209   hostonly_devicelab_tests-0-linux #824634067648: SUCCESS (cq)
I1031 13:38:48.790230       1 autoroll.go:209] autoroll.go:209   cla/google #824634066688: SUCCESS (cq)
I1031 13:38:48.790301       1 autoroll.go:209] autoroll.go:209   add_to_app_tests-macos #824634066752: SUCCESS (cq)
I1031 13:38:48.790326       1 autoroll.go:209] autoroll.go:209   web_tests-5-linux #824634067520: SUCCESS (cq)
I1031 13:38:48.790349       1 autoroll.go:209] autoroll.go:209   framework_tests-widgets-macos #824634067872: SUCCESS (cq)
I1031 13:38:48.790406       1 autoroll.go:209] autoroll.go:209   customer_testing-linux #824634066816: SUCCESS (cq)
I1031 13:38:48.790446       1 autoroll.go:209] autoroll.go:209   web_tests-2-linux #824634066944: SUCCESS (cq)
I1031 13:38:48.790510       1 autoroll.go:209] autoroll.go:209   web_tests-4-linux #824634067056: SUCCESS (cq)
I1031 13:38:48.790551       1 autoroll.go:209] autoroll.go:209   tool_tests-integration-linux #824634067104: SUCCESS (cq)
I1031 13:38:48.790621       1 autoroll.go:209] autoroll.go:209   framework_tests-libraries-linux #824634067248: SUCCESS (cq)
I1031 13:38:48.790662       1 autoroll.go:209] autoroll.go:209   tool_tests-commands-linux #824634067392: SUCCESS (cq)
I1031 13:38:48.790709       1 autoroll.go:209] autoroll.go:209   customer_testing-windows #824634067552: SUCCESS (cq)
I1031 13:38:48.790732       1 autoroll.go:209] autoroll.go:209   framework_tests-misc-linux #824634067584: SUCCESS (cq)
I1031 13:38:48.790754       1 autoroll.go:209] autoroll.go:209   web_tests-1-linux #824634067680: SUCCESS (cq)
I1031 13:38:48.790806       1 autoroll.go:209] autoroll.go:209   hostonly_devicelab_tests-2-macos #824634067728: SUCCESS (cq)
I1031 13:38:48.790829       1 autoroll.go:209] autoroll.go:209   web_tests-3-linux #824634067008: SUCCESS (cq)
I1031 13:38:48.790851       1 autoroll.go:209] autoroll.go:209   hostonly_devicelab_tests-0-macos #824634067296: SUCCESS (cq)
I1031 13:38:48.790898       1 autoroll.go:209] autoroll.go:209   framework_tests-libraries-macos #824634067360: SUCCESS (cq)
I1031 13:38:48.790922       1 autoroll.go:209] autoroll.go:209   tool_tests-integration-windows #824634066784:  (cq)
I1031 13:38:48.790943       1 autoroll.go:209] autoroll.go:209   framework_tests-misc-macos #824634067200: SUCCESS (cq)
I1031 13:38:48.790995       1 autoroll.go:209] autoroll.go:209   tool_tests-integration-macos #824634067920: SUCCESS (cq)
I1031 13:38:48.791018       1 autoroll.go:209] autoroll.go:209   web_tests-7_last-linux #824634067328: SUCCESS (cq)
I1031 13:38:48.791040       1 autoroll.go:209] autoroll.go:209   web_tests-0-linux #824634067616: SUCCESS (cq)
I1031 13:38:48.791107       1 autoroll.go:209] autoroll.go:209   framework_tests-widgets-linux #824634067792: SUCCESS (cq)
I1031 13:38:48.791138       1 autoroll.go:209] autoroll.go:209   customer_testing-macos #824634066848: SUCCESS (cq)
I1031 13:38:48.791189       1 autoroll.go:209] autoroll.go:209   analyze-linux #824634066880: SUCCESS (cq)
I1031 13:38:48.791235       1 autoroll.go:209] autoroll.go:209   web_tests-6-linux #824634067152: SUCCESS (cq)
I1031 13:38:48.791286       1 autoroll.go:209] autoroll.go:209   tool_tests-general-macos #824634067424: SUCCESS (cq)
I1031 13:38:48.791313       1 autoroll.go:209] autoroll.go:209   tool_tests-general-linux #824634067488: SUCCESS (cq)


Does anybody know what is missing and why it would be missing?

Reply:
For rollers into engine.git, this was much easier to tell: https://github.com/flutter/engine/pull/13450

The autoroller sees 9 checks which is the same as the checks listed on the PR but without WIP and it does not count each check twice. I am assuming the roller into flutter.git does the same.

Reply:
FYI: Went ahead and updated the hardcoded num of checks in all flutter github rollers (till we have a better solution): https://skia-review.googlesource.com/c/buildbot/+/251957

Reply:
Maybe the API is paged?

Reply:
Does not look like it and the results appear to be very consistent with what I described above: No WIP and no duplicates.

Reply:
Also, you will not need this script (at least for the autoroller) if you do https://github.com/flutter/flutter/issues/43672#issuecomment-548541907